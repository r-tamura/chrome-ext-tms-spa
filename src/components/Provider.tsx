import React from "react";
import { ThemeProvider } from "styled-components";
import { telema } from "~/styles/theme";

export const AppThemeProvider: React.SFC<{ children: JSX.Element }> = ({
  children
}) => <ThemeProvider theme={telema}>{children}</ThemeProvider>;
