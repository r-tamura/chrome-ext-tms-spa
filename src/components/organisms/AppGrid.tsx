import styled from "styled-components";

export const AppGrid = styled.div`
  display: grid;
  /* header content (contentの高さは view height - headerの高さ*/
  grid-template-rows: 50px minmax(calc(100vh - 50px), auto);
  /* grid-template-rows: 50px 1fr; */
  grid-template-columns: 200px 1fr;

  & > .app-header {
    grid-column: 1 / 2 span;
  }

  & > .app-content-full {
    grid-column: 1 / 2 span;
  }

  & > .app-navbar {
    grid-column: 1 / 1 span;
  }

  & > .app-content-with-navbar {
    grid-column: 2 / 1 span;
  }
`;
