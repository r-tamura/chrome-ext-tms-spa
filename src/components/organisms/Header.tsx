import React, { FC, SyntheticEvent, useState } from "react";
import Menu, { MenuItem } from "~/components/Menu";
import styled from "styled-components";
import { lighten } from "polished";
import { HIGH } from "~/styles/zIndex";
import { STANDARD } from "~/styles/boxShadow";
import { XL } from "~/styles/font";
import { Theme } from "~/styles/theme";
import { useUser } from "~/stores/hooks";

interface IHeaderUserProps {
  onLogout: () => void;
}

const HeaderUser: FC<IHeaderUserProps> = ({ onLogout, children }) => {
  const [anchorElm, setAnchorElm] = useState(null);

  const menuOpened = anchorElm != null;
  return (
    <UserMenu>
      <ClickableSpan id={"header-user-button"} onClick={handleClick}>
        {children}
      </ClickableSpan>
      <Menu title={"User Menu"} onClose={handleClose} open={menuOpened}>
        <MenuItem id={"logout"} button onClick={handleLogout}>
          Logout
        </MenuItem>
      </Menu>
    </UserMenu>
  );

  function handleClick(e: SyntheticEvent<Element>) {
    setAnchorElm(e.currentTarget);
  }

  function handleClose(_: MouseEvent) {
    setAnchorElm(null);
  }

  function handleLogout(e: SyntheticEvent<Element, MouseEvent>) {
    handleClose(e.nativeEvent);
    onLogout();
  }
};

/**
 * アプリケーションヘッダーコンポーネント
 */
export const AppHeader: FC = () => {
  const { logout, name: username } = useUser();
  return (
    <Root className="app-header">
      <CompanyLogoAnchor href="http://www.telema.jp/">
        <img src="https://www.telema.co.jp/images/logo_sp.svg" alt="logo" />
      </CompanyLogoAnchor>
      <HeaderUser onLogout={logout}>{username}</HeaderUser>
    </Root>
  );
};

const ClickableSpan = styled.span`
  &:hover {
    cursor: pointer;
    text-decoration: none;
  }
`;

const CompanyLogoAnchor = styled.a`
  display: block;
  width: 40px;
  &:focus,
  &:hover {
    opacity: 0.8;
    text-decoration: none;
  }
`;

const Root = styled.header`
  display: flex;
  align-items: center;
  flex-direction: row;
  padding: 0 5px;
  background-color: ${({ theme }) => theme.primaryReverse};
  font-size: ${XL};
  font-weight: bolder;
  line-height: 50px;

  z-index: ${HIGH};
  box-shadow: ${STANDARD};
`;

const UserMenu = styled.div`
  display: flex;
  align-items: center;
  margin-left: auto;
  color: ${({ theme }) => theme.textMain};

  .person-icon {
    background-image: url(chrome-extension://odbcdehjjdjjnpenjbblpmkkibggblpe/chrome/assets/images/person.svg);
    height: 50%;
    width: 25px;
    display: block;
  }

  &:hover {
    color: ${({ theme }: { theme: Theme }) => lighten(0.3, theme.textMain)};
  }
`;
