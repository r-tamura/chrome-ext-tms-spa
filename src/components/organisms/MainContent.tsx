import styled from "styled-components";

export const MainContent = styled.div`
  padding: 10px;
`;
