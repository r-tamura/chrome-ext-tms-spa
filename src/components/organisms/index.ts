export * from "./AppGrid";
export * from "./MainContent";
export * from "./Header";
export * from "./TransExpenseHistory";
export * from "./TransExpenseTemplates";
export * from "./Nav";
