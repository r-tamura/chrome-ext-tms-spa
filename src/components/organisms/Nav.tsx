import React from "react";
import { NavItem } from "~/components/molecules";
import styled from "styled-components";
import { MID } from "~/styles/zIndex";
import { STANDARD } from "~/styles/boxShadow";
import { ThemeProps } from "~/styles/theme";

interface INavProps extends React.Props<{}> {
  path: string;
}

export const Nav: React.SFC<INavProps> = ({ path }) => {
  const pathname = path.replace(/^\/([^/]+).*$/, "$1");
  return (
    <NavStyled id="side-nav-bar">
      <ul>
        {/* リンク一覧 */}
        <li>
          <NavItem
            to={"transportation"}
            selected={"transportation" === pathname}
          >
            交通費
          </NavItem>
        </li>
        <li>
          <NavItem to={"attendance"} selected={"attendance" === pathname}>
            勤怠管理
          </NavItem>
        </li>
        {/* TODO: 資源管理は未実装 */}
        <li>
          <NavItem
            to={"resource"}
            disabled={true}
            selected={"resource" === pathname}
          >
            資源管理
          </NavItem>
        </li>
      </ul>
    </NavStyled>
  );
};

const NavStyled = styled.nav<ThemeProps>`
  background-color: ${({ theme }) => theme.primary};
  z-index: ${MID};
  box-shadow: ${STANDARD};
`;
