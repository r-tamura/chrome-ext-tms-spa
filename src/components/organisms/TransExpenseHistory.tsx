import React, { useState } from "react";
import { find, propEq } from "ramda";
import {
  Button,
  ButtonGroup,
  Panel,
  RequiredInput,
  Modal
} from "~/components/atoms";
import { AddIcon } from "~/components/Icon";
import { CURRENCY } from "~/helpers/_const";
import {
  Project,
  Usage,
  Objective,
  TransExpenseView,
  TransExpenseTemplateView,
  ExpenseId,
  TransExpenseCreateRequest,
  TransExpenseUpdateRequest,
  TransExpense
} from "~/types";

import { Table, Tr, Td } from "~/components/atoms/Table";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { faEdit, faTrash } from "@fortawesome/free-solid-svg-icons";
import { useModal } from "~/components/hooks";
import useForm from "react-hook-form";
import { SelectBox } from "../SelectBox";

interface IProps {
  expenses: TransExpenseView[];
  templates: TransExpenseTemplateView[];
  projects: Project[];
  usages: Usage[];
  objectives: Objective[];
  deleteExpense: (expenseId: ExpenseId) => void;
  createExpense: (expense: TransExpenseCreateRequest) => void;
  updateExpense: (expense: TransExpenseUpdateRequest) => void;
  createExpenseFromTemplate: (templateId: string, strdate: string) => void;
}

interface IState {
  selected?: ExpenseId; // 選択中の交通費ID
  useTemplate?: boolean;
}

/**
 * 登録済み交通費リストセクション
 */
export const TransExpenseHistory: React.FC<IProps> = props => {
  const { isOpen: isModalOpen, open, close } = useModal(false);
  const [state, setState] = useState(getInitState(props));

  const { expenses, deleteExpense } = props;
  const { useTemplate } = state;
  return (
    <div>
      {expenses.length > 0 ? <Expenses /> : <p>交通費が登録されていません。</p>}
      <ButtonGroup>
        <Button
          variant={"contained"}
          color={"secondary"}
          onClick={openCreateFromTemplateModal}
        >
          テンプレートから新規登録
        </Button>
        <Button
          id={"create-trans-expense"}
          variant={"circle"}
          color={"primary"}
          onClick={openCreateModal}
        >
          <AddIcon />
        </Button>
      </ButtonGroup>

      {/* 編集モーダル */}
      <EditorModal />
    </div>
  );

  function Expenses() {
    const header = [
      "#",
      "利用日",
      "プロジェクト",
      "利用区分",
      "目的",
      "金額",
      "",
      ""
    ];
    return (
      <div>
        <Table header={header}>
          {expenses.map((e, i) => (
            <Tr key={(() => e.expenseId)()} hover>
              <Td className="record-id">{i + 1}</Td>
              <Td>{e.strdate}</Td>
              <Td>{e.project.name}</Td>
              <Td>{e.usage.name}</Td>
              <Td>{e.objective.name}</Td>
              <Td>
                <strong>{CURRENCY.format(e.cost)}</strong>
              </Td>
              <Td>
                <Button
                  variant="icon"
                  color="primary"
                  onClick={() => openEditModal(e.expenseId)}
                  title="編集"
                >
                  <FontAwesomeIcon icon={faEdit} />
                </Button>
              </Td>
              <Td>
                <Button
                  variant="icon"
                  color="danger"
                  onClick={() => deleteExpense(e.expenseId.toString())}
                  title="削除"
                >
                  <FontAwesomeIcon icon={faTrash} />
                </Button>
              </Td>
            </Tr>
          ))}
        </Table>
      </div>
    );
  }

  /*
   * 交通費新規作成/編集モーダルをレンダーします
   */
  function EditorModal() {
    type FormData = Record<keyof TransExpense, string>;
    const {
      register,
      handleSubmit,
      clearError,
      errors,
      setError,
      formState
    } = useForm<FormData>();
    const {
      expenses,
      projects,
      usages,
      objectives,
      createExpense,
      updateExpense
    } = props;
    const { selected } = state;

    function handleCancelClick(e: React.MouseEvent) {
      e.preventDefault();
      e.stopPropagation();
      closeModal();
    }

    function onSubmit({
      strdate,
      projectId,
      usageId,
      objectiveId,
      customer,
      from,
      to,
      cost: costStr
    }: FormData) {
      let cost;
      cost = Number.parseInt(costStr, 10);
      if (Number.isNaN(cost)) {
        setError(
          "cost",
          "NaN",
          "有効な金額ではありません、正の整数値を入力してください"
        );
        return;
      }
      if (cost < 0) {
        setError("cost", "NaN", "正の整数値を入力してください");
        return;
      }

      if (selected) {
        updateExpense({
          expenseId,
          strdate,
          projectId,
          usageId,
          objectiveId,
          customer,
          from,
          to,
          cost
        });
      } else {
        createExpense({
          strdate,
          projectId,
          usageId,
          objectiveId,
          customer,
          from,
          to,
          cost
        });
      }
      closeModal();
    }

    const defaultExpense: TransExpenseView = {
      expenseId: null,
      strdate: "",
      project: projects[0],
      usage: usages[0],
      objective: objectives[0],
      customer: "",
      from: "",
      to: "",
      cost: 0
    };

    const {
      expenseId,
      strdate,
      project,
      usage,
      objective,
      customer,
      from,
      to,
      cost
    }: TransExpenseView =
      find(propEq("expenseId", selected), expenses) || defaultExpense;
    const costStr = cost.toString();
    return (
      <Modal isOpen={isModalOpen} onRequestClose={() => close()}>
        <Panel>
          <form onSubmit={handleSubmit(onSubmit)}>
            <RequiredInput
              name={"strdate"}
              label={"日付"}
              placeholder={"ex: 20190401"}
              error={errors.strdate}
              defaultValue={strdate}
              register={register}
              clearError={clearError}
            />
            <SelectBox
              name="projectId"
              label="プロジェクトコード *"
              defaultValue={project.projectId}
              options={{
                items: projects,
                value: (project: Project) => project.projectId,
                label: (project: Project) => project.name
              }}
              ref={register({ required: true })}
            />
            <SelectBox
              name="usageId"
              label="利用区分 *"
              defaultValue={usage.usageId}
              options={{
                items: usages,
                value: (usage: Usage) => usage.usageId,
                label: (usage: Usage) => usage.name
              }}
              ref={register({ required: true })}
            />
            <SelectBox
              name="objectiveId"
              label="目的コード *"
              defaultValue={objective.objectiveId}
              options={{
                items: objectives,
                value: (objective: Objective) => objective.objectiveId,
                label: (objective: Objective) => objective.name
              }}
              ref={register({ required: true })}
            />
            <RequiredInput
              name={"customer"}
              label={"顧客先"}
              placeholder={"ex: XXXX 株式会社"}
              error={errors.customer}
              defaultValue={customer}
              register={register}
              clearError={clearError}
            />
            <RequiredInput
              name={"from"}
              label={"乗車地"}
              placeholder={"ex: JR 千葉駅"}
              error={errors.from}
              defaultValue={from}
              register={register}
              clearError={clearError}
            />
            <RequiredInput
              name={"to"}
              label={"降車地"}
              placeholder={"ex: JR 東京駅"}
              error={errors.to}
              defaultValue={to}
              register={register}
              clearError={clearError}
            />
            <RequiredInput
              name={"cost"}
              label={"金額"}
              placeholder={"ex: 500"}
              error={errors.cost}
              defaultValue={costStr}
              register={register}
              clearError={clearError}
            />
            <ButtonGroup>
              <Button variant="contained" onClick={handleCancelClick}>
                キャンセル
              </Button>
              <Button
                variant="contained"
                color="primary"
                disabled={!formState.isValid}
              >
                {expenseId ? "更新" : "作成"}
              </Button>
            </ButtonGroup>
          </form>
        </Panel>
      </Modal>
    );
  }

  function renderCreateFromTemplateModal() {
    const { templates, createExpenseFromTemplate } = props;
    return (
      // <Modal
      //   isOpen={isModalOpen}
      //   formItems={[
      //     {
      //       type: FormItemType.TEXT,
      //       name: "strdate",
      //       label: "日付",
      //       value: ""
      //     },
      //     {
      //       type: FormItemType.SELECT,
      //       name: "templateId",
      //       label: "テンプレート名",
      //       value: "",
      //       options: {
      //         items: templates,
      //         valueKey: "templateId",
      //         labelKey: "templateName"
      //       }
      //     }
      //   ]}
      //   onClose={closeModal}
      //   onOKClick={({
      //     templateId,
      //     strdate
      //   }: {
      //     templateId: string;
      //     strdate: string;
      //   }) => createExpenseFromTemplate(templateId, strdate)}
      // />
      <div> history template</div>
    );
  }

  function getInitState(props: IProps): IState {
    return { useTemplate: false, selected: undefined };
  }

  /* モーダルクローズ */
  function closeModal() {
    setState({
      useTemplate: false,
      selected: null
    });
    close();
  }

  /* 編集モーダル */
  function openEditModal(expenseId: ExpenseId) {
    setState({
      useTemplate: false,
      selected: expenseId
    });
    open();
  }

  /* 新規登録モーダル */
  function openCreateModal() {
    setState({
      useTemplate: false,
      selected: null
    });
    open();
  }

  /* テンプレートから新規登録モーダル */
  function openCreateFromTemplateModal() {
    setState({
      useTemplate: true,
      selected: null
    });
    open();
  }
};
