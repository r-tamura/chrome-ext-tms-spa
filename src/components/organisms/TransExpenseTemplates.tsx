import React, { useState } from "react";
import useForm from "react-hook-form";
import { find, propEq } from "ramda";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { faTrash, faEdit } from "@fortawesome/free-solid-svg-icons";
import { Modal } from "~/components/atoms/Modal";
import { Button, ButtonGroup, Panel, Input } from "~/components/atoms";
import { AddIcon } from "~/components/Icon";
import { CURRENCY } from "~/helpers/_const";
import {
  Project,
  Usage,
  Objective,
  TransExpenseTemplateView,
  TransExpenseTemplateCreateRequest,
  TransExpenseTemplateUpdateRequest,
  TransExpenseTemplate,
  TemplateId
} from "~/types";
import { Table, Tr, Td } from "~/components/atoms/Table";
import { useModal } from "~/components/hooks";
import { SelectBox } from "../SelectBox";
import { ReactHookFormError } from "react-hook-form/dist/types";

interface IProps {
  templates: TransExpenseTemplateView[];
  projects: Project[];
  usages: Usage[];
  objectives: Objective[];
  createExpenseTemplate: (template: TransExpenseTemplateCreateRequest) => void;
  deleteExpenseTemplate: (templateId: string) => void;
  updateExpenseTemplate: (template: TransExpenseTemplateUpdateRequest) => void;
}

/**
 * 交通費テンプレートコンポーネント
 */
export const TransExpenseTemplates: React.FC<IProps> = ({
  templates,
  projects,
  usages,
  objectives,
  createExpenseTemplate,
  updateExpenseTemplate,
  deleteExpenseTemplate
}) => {
  const { isOpen: isModalOpen, open, close } = useModal(false);
  const [selected, setSelected] = useState<TemplateId | null>(null);
  return (
    <div>
      {/* テンプレート一覧 */}
      {templates.length > 0 ? (
        <TemplatesTable />
      ) : (
        <p>交通費テンプレートがブラウザに登録されていません。登録しますか？</p>
      )}
      {/* 新規登録ボタン */}
      <ButtonGroup>
        <Button
          id={"create-trans-expense-template"}
          color={"primary"}
          variant={"circle"}
          onClick={openCreateModal}
        >
          <AddIcon />
        </Button>
      </ButtonGroup>

      {/* 編集モーダル */}
      <EditorModal />
    </div>
  );

  function TemplatesTable() {
    const header = [
      "#",
      "テンプレート名",
      "プロジェクト",
      "利用区分",
      "目的",
      "金額",
      "",
      ""
    ];

    function createHandleDelete(templateId: TemplateId) {
      return () => deleteExpenseTemplate(templateId);
    }
    return (
      <div>
        <Table header={header}>
          {templates.map((t, i) => (
            <Tr key={t.templateId} hover>
              <Td className="record-id">{i + 1}</Td>
              <Td>{t.templateName}</Td>
              <Td>{t.project.name}</Td>
              <Td>{t.usage.name}</Td>
              <Td>{t.objective.name}</Td>
              <Td>{CURRENCY.format(t.cost)}</Td>
              <Td>
                <Button
                  onClick={() => openEditModal(t.templateId)}
                  variant="icon"
                  color="primary"
                  title="編集"
                >
                  <FontAwesomeIcon icon={faEdit} />
                </Button>
              </Td>
              <Td>
                <Button
                  variant="icon"
                  color="danger"
                  title="削除"
                  onClick={createHandleDelete(t.templateId)}
                >
                  <FontAwesomeIcon icon={faTrash} />
                </Button>
              </Td>
            </Tr>
          ))}
        </Table>
      </div>
    );
  }

  /*
   * 交通費新規作成/編集モーダルをレンダーします
   */
  function EditorModal() {
    type FormData = Record<keyof TransExpenseTemplate, string>;
    const {
      register,
      handleSubmit,
      errors,
      setError,
      formState,
      clearError
    } = useForm<FormData>();
    function onSubmit({
      templateName,
      projectId,
      usageId,
      objectiveId,
      customer,
      from,
      to,
      cost
    }: FormData) {
      let costNumber;
      try {
        costNumber = Number.parseInt(cost, 10);
        if (costNumber < 0) {
          setError("cost", "NaN", "正の整数値を入力してください");
          return;
        }
      } catch (e) {
        setError("cost", "NaN", "正の整数値を入力してください");
        return;
      }

      if (selected) {
        // 更新
        updateExpenseTemplate({
          templateId: find<TransExpenseTemplateView>(
            propEq("templateId", selected),
            templates
          ).templateId,
          templateName,
          projectId,
          usageId,
          objectiveId,
          customer,
          from,
          to,
          cost: costNumber
        });
      } else {
        // 新規作成
        createExpenseTemplate({
          templateName,
          projectId,
          usageId,
          objectiveId,
          customer,
          from,
          to,
          cost: costNumber
        });
      }
      close();
    }

    function handleCancelClick(event: React.MouseEvent) {
      event.stopPropagation();
      event.preventDefault();
      close();
    }

    const defaultTemplate: TransExpenseTemplateView = {
      templateId: "",
      templateName: "",
      project: null,
      usage: null,
      objective: null,
      customer: "",
      from: "",
      to: "",
      cost: 0
    };
    const {
      templateName,
      project,
      usage,
      objective,
      customer,
      from,
      to,
      cost
    } = selected
      ? find(propEq("templateId", selected), templates)
      : defaultTemplate;
    const costStr = cost === 0 ? "" : cost.toString();

    function RequiredInput({
      label,
      ...otherProps
    }: {
      name: keyof FormData;
      label: string;
      placeholder: string;
      defaultValue: string;
      error: ReactHookFormError;
    }) {
      return (
        <Input
          {...otherProps}
          error={!!otherProps.error}
          label={label + " *"}
          ref={register({ required: `${label}は必須入力です` })}
          onFocus={() => clearError(otherProps.name)}
        />
      );
    }

    return (
      <Modal isOpen={isModalOpen} onRequestClose={handleModalClose}>
        <Panel>
          <form onSubmit={handleSubmit(onSubmit)}>
            <RequiredInput
              name="templateName"
              label="テンプレート名"
              placeholder="ex: 定期代(JR千葉駅)"
              defaultValue={templateName}
              error={errors.templateName}
            />
            <SelectBox
              name="projectId"
              label="プロジェクトコード *"
              defaultValue={project ? project.projectId : ""}
              options={{
                items: projects,
                value: (project: Project) => project.projectId,
                label: (project: Project) => project.name
              }}
              ref={register({ required: true })}
            />
            <SelectBox
              name="usageId"
              label="利用区分 *"
              defaultValue={usage ? usage.usageId : ""}
              options={{
                items: usages,
                value: (usage: Usage) => usage.usageId,
                label: (usage: Usage) => usage.name
              }}
              ref={register({ required: true })}
            />
            <SelectBox
              name="objectiveId"
              label="目的コード *"
              defaultValue={objective ? objective.objectiveId : ""}
              options={{
                items: objectives,
                value: (objective: Objective) => objective.objectiveId,
                label: (objective: Objective) => objective.name
              }}
              ref={register({ required: true })}
            />
            <RequiredInput
              name="customer"
              label="顧客先"
              placeholder="ex: XX株式会社"
              defaultValue={customer}
              error={errors.customer}
            />
            <RequiredInput
              name="from"
              label="乗車地"
              placeholder="ex: JR千葉駅"
              defaultValue={from}
              error={errors.from}
            />
            <RequiredInput
              name="to"
              label="降車地"
              placeholder="ex: JR東京駅"
              defaultValue={to}
              error={errors.to}
            />
            <RequiredInput
              name="cost"
              label="金額"
              placeholder="500"
              defaultValue={costStr}
              error={errors.cost}
            />
            <ButtonGroup>
              <Button variant="contained" onClick={handleCancelClick}>
                キャンセル
              </Button>
              <Button
                variant="contained"
                color="primary"
                disabled={!formState.isValid}
              >
                作成/更新
              </Button>
            </ButtonGroup>
          </form>
        </Panel>
      </Modal>
    );
  }

  function handleModalClose() {
    close();
  }

  /* 編集モーダルオープン */
  function openEditModal(templateId: string) {
    setSelected(templateId);
    open();
  }

  /* 新規登録ボタンクリックイベント */
  function openCreateModal(e: React.MouseEvent<HTMLButtonElement>) {
    open();
    setSelected("");
  }
};
