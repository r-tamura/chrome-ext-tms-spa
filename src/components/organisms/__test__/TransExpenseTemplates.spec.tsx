import React from "react";
import { mount, shallow, ReactWrapper } from "enzyme";
import { TransExpenseTemplates } from "~/components/organisms/TransExpenseTemplates";
import { AppThemeProvider } from "~/components/Provider";
import toJson from "enzyme-to-json";
import { TransExpenseTemplateView } from "~/types";
import { wait, setValue } from "~/components/test_util";

function clickCreateButton(wrapper: ReactWrapper) {
  wrapper
    .find("button")
    .last()
    .simulate("click");
}

function clickEditButton(wrapper: ReactWrapper, index: number = 0) {
  wrapper
    .find("tbody tr")
    .at(index)
    .find("button")
    .first()
    .simulate("click");
}

describe("<TransExpenseTemplates />", () => {
  it("shallow snap shot", () => {
    const spyCreate = jest.fn();
    const spyUpdate = jest.fn();
    const spyDelete = jest.fn();
    const $ = shallow(
      <AppThemeProvider>
        <TransExpenseTemplates
          templates={[]}
          projects={[]}
          usages={[]}
          objectives={[]}
          createExpenseTemplate={spyCreate}
          updateExpenseTemplate={spyUpdate}
          deleteExpenseTemplate={spyDelete}
        />
      </AppThemeProvider>
    );

    expect(toJson($)).toMatchSnapshot("");
  });

  describe("mount with no templates", () => {
    const spyCreate = jest.fn();
    const spyUpdate = jest.fn();
    const spyDelete = jest.fn();
    const $ = mount(
      <AppThemeProvider>
        <TransExpenseTemplates
          templates={[]}
          projects={[]}
          usages={[]}
          objectives={[]}
          createExpenseTemplate={spyCreate}
          updateExpenseTemplate={spyUpdate}
          deleteExpenseTemplate={spyDelete}
        />
      </AppThemeProvider>
    );
    it("交通費がテンプレートに未登録である旨を表示する", () => {
      expect(
        $.text().includes(
          "交通費テンプレートがブラウザに登録されていません。登録しますか？"
        )
      ).toBe(true);
    });

    it("作成ボタンを押す -> テンプレート作成ダイアログを表示する", () => {
      expect($.find("form").exists()).toBe(false);
      $.find("button#create-trans-expense-template").simulate("click");
      expect($.find("form").exists()).toBe(true);
    });
  });

  describe("mount with some templates", () => {
    const spyCreate = jest.fn();
    const spyUpdate = jest.fn();
    const spyDelete = jest.fn();
    const fakeTemplates: TransExpenseTemplateView[] = [
      {
        templateId: "98bd5d97-a169-4fc1-a8a7-f1565e48f162",
        templateName: "template 1",
        customer: "TELEMA",
        project: {
          name: "",
          projectId: ""
        },
        usage: {
          name: "",
          usageId: ""
        },
        objective: {
          name: "",
          objectiveId: ""
        },
        from: "Tokyo Station",
        to: "Chiba Station",
        cost: 999
      },
      {
        templateId: "12azkz68-d769-4fc1-a8n8-b1885e48f162",
        templateName: "template 2",
        project: {
          name: "",
          projectId: ""
        },
        usage: {
          name: "",
          usageId: ""
        },
        objective: {
          name: "",
          objectiveId: ""
        },
        customer: "Dummy Company Ctd.",
        from: "Sapporo Station",
        to: "Aomori Station",
        cost: 1240
      },
      {
        templateId: "b83014cd-c6d7-478a-b133-7363b710832d",
        templateName: "template 3",
        project: {
          name: "",
          projectId: ""
        },
        usage: {
          name: "",
          usageId: ""
        },
        objective: {
          name: "",
          objectiveId: ""
        },
        customer: "Dummy Company Ctd.",
        from: "Yokohama Station",
        to: "Shinjuku Station",
        cost: 560
      }
    ];
    const $ = mount(
      <AppThemeProvider>
        <TransExpenseTemplates
          templates={fakeTemplates}
          projects={[{ name: "Project X", projectId: "201108010001" }]}
          usages={[{ name: "交通費(往復)", usageId: "3" }]}
          objectives={[{ name: "作業", objectiveId: "2" }]}
          createExpenseTemplate={spyCreate}
          updateExpenseTemplate={spyUpdate}
          deleteExpenseTemplate={spyDelete}
        />
      </AppThemeProvider>
    );

    it("指定されたテンプレート数のレコードを表示する", () => {
      expect($.find("tbody tr").length).toBe(3);
    });

    describe("作成モーダル", () => {
      beforeAll(() => {
        clickCreateButton($);
      });

      it("作成モーダルフォーム", () => {
        expect($.find("form").length).toBe(1);
      });

      describe("未入力項目ありでサブミット", () => {
        beforeAll(() => {
          $.find("form").simulate("submit");
        });

        it("テンプレート作成関数は呼ばれない", async () => {
          await wait();
          expect(spyCreate.mock.calls.length).toBe(0);
        });
      });

      describe("未入力項目なしでサブミット", () => {
        beforeAll(() => {
          setValue($, "templateName", "サンプルテンプレート");
          setValue($, "customer", "XXXX 株式会社");
          setValue($, "from", "JR 東京駅");
          setValue($, "to", "JR 名古屋駅");
          setValue($, "cost", "1000");
          $.find("form").simulate("submit");
        });

        it("テンプレート作成関数が呼ばれる", async () => {
          await wait();
          expect(spyCreate.mock.calls.length).toBe(1);
          expect(spyCreate.mock.calls[0][0]).toEqual({
            templateName: "サンプルテンプレート",
            customer: "XXXX 株式会社",
            from: "JR 東京駅",
            to: "JR 名古屋駅",
            cost: 1000,
            objectiveId: "2",
            projectId: "201108010001",
            usageId: "3"
          });
        });
      });
    });

    describe("編集モーダル", () => {
      beforeAll(() => {
        clickEditButton($);
      });

      it("編集モーダルフォーム", () => {
        expect($.find("form").length).toBe(1);
      });

      describe("未入力項目なしでサブミット", () => {
        beforeAll(() => {
          setValue($, "templateName", "サンプルテンプレート");
          setValue($, "customer", "XXXX 株式会社");
          setValue($, "from", "JR 東京駅");
          setValue($, "to", "JR 名古屋駅");
          setValue($, "cost", "1000");
          $.find("form").simulate("submit");
        });

        it("テンプレート作成関数が呼ばれる", async () => {
          await wait();
          expect(spyUpdate.mock.calls.length).toBe(1);
          expect(spyUpdate.mock.calls[0][0]).toEqual({
            templateId: "98bd5d97-a169-4fc1-a8a7-f1565e48f162",
            templateName: "サンプルテンプレート",
            customer: "XXXX 株式会社",
            from: "JR 東京駅",
            to: "JR 名古屋駅",
            cost: 1000,
            objectiveId: "2",
            projectId: "201108010001",
            usageId: "3"
          });
        });
      });
    });
  });
});
