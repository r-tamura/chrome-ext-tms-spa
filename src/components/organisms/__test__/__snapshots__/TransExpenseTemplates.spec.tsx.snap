// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TransExpenseTemplates /> shallow snap shot 1`] = `
<ThemeProvider
  theme={
    Object {
      "black": "rgb(5, 5, 5)",
      "buttonDefault": "rgba(0, 0, 0, 0.08)",
      "danger": "#D32F2F",
      "fontFamily": "\\"M PLUS Rounded 1c\\", sans-serif",
      "formBorderColor": "rgba(0, 0, 0, 0.16)",
      "formBorderColorDisabled": "rgba(0, 0, 0, 0.08)",
      "formBorderColorError": "#D32F2F",
      "formBorderColorFocused": "hsl(206, 100%, 31%)",
      "hoveredBg": "rgba(0, 0, 0, 0.03)",
      "inputPlaceHolder": "rgba(0, 0, 0, 0.09)",
      "primary": "hsl(206, 100%, 31%)",
      "primaryDark": "hsl(208, 80%, 27%)",
      "primaryHeavyDark": "hsl(208, 80%, 27%)",
      "primaryHeavyLight": "hsl(199, 91%, 64%);",
      "primaryLight": "hsl(201, 98%, 41%);",
      "primaryReverse": "hsla(0, 0%, 99%, 1);",
      "primaryReverseDark": "hsla(0, 0%, 99%, 0.5);",
      "secondary": "#689F38",
      "textGray": "hsla(0, 5%, 5%, .25)",
      "textMain": "hsl(0, 5%, 5%)",
      "white": "rgb(253, 253, 253)",
    }
  }
>
  <Component
    createExpenseTemplate={[MockFunction]}
    deleteExpenseTemplate={[MockFunction]}
    objectives={Array []}
    projects={Array []}
    templates={Array []}
    updateExpenseTemplate={[MockFunction]}
    usages={Array []}
  />
</ThemeProvider>
`;
